staging/src/k8s.io/apiserver/pkg/util/flowcontrol/watch_tracker.go

```diff
@@ -91,16 +150,31 @@ func (w *watchTracker) RegisterWatch(requestInfo *request.RequestInfo) ForgetWat
 
 	w.lock.Lock()
 	defer w.lock.Unlock()
-	w.watchCount[*identifier]++
-	return w.forgetWatch(identifier)
+	w.updateIndexLocked(identifier, index, 1)
+	return w.forgetWatch(identifier, index)
+}
+
+func (w *watchTracker) updateIndexLocked(identifier *watchIdentifier, index *indexValue, incr int) {
+	if index == nil {
+		w.watchCount[*identifier]+=incr
+	} else {
+		// TODO(wojtek-t): For now we estimate the number of watchers as sum
+		// of those which don't specify the index selector and those for an
+		// empty value (we assume that number of those for empty value is not
+		// smaller than for any other value). That's not always true, but in
```

Comment:
I think the assumption being made here is rather that the count for not-unset non-empty value stays tiny in absolute terms.

Reply:
Not exactly (and what you're saying is also not true in practice - e.g. each kubelet is watching for its own pods, which means that there are N such watches with N being number of nodes).

The observation is that for a given object create/update/delete the amount of work (also filtering given index implementaion in watchcache) is proportional to:
- watches interested in all objects (those that are "unset" here)
- watches interssted in the value coming from that object

So we're basically assuming (which seem to be true in practice too), that the difference between "unset + my-value" is small comparing to "unset + empty".

Tried to clarify the comment.